[BUG] Compiler warning reported by clang-19 in zipf implementation

```
  [289/321] /opt/homebrew/opt/llvm/bin/clang -Inumpy/random/mtrand.cpython-313t-darwin.so.p -Inumpy/random -I../numpy/random -I../numpy/random/src -Inumpy/_core -I../numpy/_core -Inumpy/_core/include -I../numpy/_core/include -I../numpy/_core/src/common -Inumpy -I/Users/goldbaum/.pyenv/versions/3.13t-dev-tsan/include/python3.13t -I/Users/goldbaum/Documents/numpy/build/meson_cpu -fvisibility=hidden -fdiagnostics-color=always -fsanitize=thread -DNDEBUG -Wall -Winvalid-pch -std=c11 -O3 -fno-strict-aliasing -ftrapping-math -DNPY_HAVE_CLANG_FPSTRICT -DNPY_HAVE_NEON_VFPV4 -DNPY_HAVE_NEON_FP16 -DNPY_HAVE_NEON -DNPY_HAVE_ASIMD -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -DNPY_NO_DEPRECATED_API=0 -DNP_RANDOM_LEGACY=1 -MD -MQ numpy/random/mtrand.cpython-313t-darwin.so.p/src_distributions_distributions.c.o -MF numpy/random/mtrand.cpython-313t-darwin.so.p/src_distributions_distributions.c.o.d -o numpy/random/mtrand.cpython-313t-darwin.so.p/src_distributions_distributions.c.o -c ../numpy/random/src/distributions/distributions.c
  ../numpy/random/src/distributions/distributions.c:1026:14: warning: implicit conversion from 'long' to 'double' changes value from 9223372036854775807 to 9223372036854775808 [-Wimplicit-const-int-float-conversion]
   1026 |   Umin = pow(RAND_INT_MAX, -am1);
        |          ~~~ ^~~~~~~~~~~~
  ../numpy/_core/include/numpy/random/distributions.h:25:22: note: expanded from macro 'RAND_INT_MAX'
     25 | #define RAND_INT_MAX LONG_MAX
        |                      ^~~~~~~~
  /opt/homebrew/Cellar/llvm/19.1.6/lib/clang/19/include/limits.h:51:19: note: expanded from macro 'LONG_MAX'
     51 | #define LONG_MAX  __LONG_MAX__
        |                   ^~~~~~~~~~~~
  <built-in>:64:22: note: expanded from macro '__LONG_MAX__'
     64 | #define __LONG_MAX__ 9223372036854775807L
        |                      ^~~~~~~~~~~~~~~~~~~~
  1 warning generated.
```

Originates in this line:

https://github.com/numpy/numpy/blob/206f697a702968c6815c3fae48370878afd3a837/numpy/random/src/distributions/distributions.c#L1026

No idea if this is finding a real edge case that could lead to an error in the calculation, since `Umin` is calculated based on a lossy conversion to double.

--- a/numpy/random/_common.pyx
+++ b/numpy/random/_common.pyx
@@ -918,31 +918,33 @@ cdef object disc(void *func, void *state, object size, object lock,
         else:
             raise NotImplementedError("No vector path available")
 
+    # At this point, we know is_scalar is True.
+
     if narg_double > 0:
         _da = PyFloat_AsDouble(a)
-        if a_constraint != CONS_NONE and is_scalar:
+        if a_constraint != CONS_NONE:
             check_constraint(_da, a_name, a_constraint)
 
         if narg_double > 1:
             _db = PyFloat_AsDouble(b)
-            if b_constraint != CONS_NONE and is_scalar:
+            if b_constraint != CONS_NONE:
                 check_constraint(_db, b_name, b_constraint)
         elif narg_int64 == 1:
             _ib = <int64_t>b
-            if b_constraint != CONS_NONE and is_scalar:
+            if b_constraint != CONS_NONE:
                 check_constraint(<double>_ib, b_name, b_constraint)
     else:
         if narg_int64 > 0:
             _ia = <int64_t>a
-            if a_constraint != CONS_NONE and is_scalar:
+            if a_constraint != CONS_NONE:
                 check_constraint(<double>_ia, a_name, a_constraint)
         if narg_int64 > 1:
             _ib = <int64_t>b
-            if b_constraint != CONS_NONE and is_scalar:
+            if b_constraint != CONS_NONE:
                 check_constraint(<double>_ib, b_name, b_constraint)
         if narg_int64 > 2:
             _ic = <int64_t>c
-            if c_constraint != CONS_NONE and is_scalar:
+            if c_constraint != CONS_NONE:
                 check_constraint(<double>_ic, c_name, c_constraint)
 
     if size is None:

--- a/numpy/random/_common.pyx
+++ b/numpy/random/_common.pyx
@@ -467,7 +467,7 @@ cdef int check_constraint(double val, object name, constraint_type cons) except 
         # Note: Assume value is integral (double of LONG_MAX should work out)
         if val < 0:
             raise ValueError(name + " < 0")
-        elif val > LONG_MAX:
+        elif val > <double> LONG_MAX:
             raise ValueError(
                     name + " is out of bounds for long, consider using "
                     "the new generator API for 64bit integers.")

--- a/numpy/random/src/distributions/distributions.c
+++ b/numpy/random/src/distributions/distributions.c
@@ -1023,7 +1023,7 @@ RAND_INT_TYPE random_zipf(bitgen_t *bitgen_state, double a) {
    * Values below Umin would result in X being rejected because it is too
    * large, so there is no point in including them in the distribution of U.
    */
-  Umin = pow(RAND_INT_MAX, -am1);
+  Umin = pow((double) RAND_INT_MAX, -am1);
   while (1) {
     double U01, T, U, V, X;
 

